### PR TITLE
[TLS 1.3] Consolidate the Session Manager

### PR DESCRIPTION
**Pull request by reneme**
### Pull Request Dependencies

* ~~#3165~~
* ~~#3201~~
* ~~#3244~~

### Strong Types

Those are the first code changes that make use of the strong-type helper introduced in #3165. Particularly for `Session_ID` and `Session_Ticket`, which are otherwise simply `std::vector<uint8_t>` containers.

### `class Session`

This is now a dumb data-sink that can (un)marshal itself (and knows how to encrypt/decrypt itself). In particular it doesn't have a notion about its own ID or ticket anymore. These are the realm of the `Session_Manager`.

### `struct Session_Handle`

A descriptor for a session. As in the example above: It contains either a (TLS 1.2-ish) ID or a ticket (TLS 1.2 or 1.3) but never both.  These invariants are ensured by the class. The `Session_Manager` emits a `Session_Handle` for new sessions and spits them out again given a `Session_Handle`. It can also store a `Session` along with an existing session handle.

### `class Session_Manager`

I put some documentation in the header, so won't repeat that here. Main point: There is a sharp distinction between APIs useful for the server and the client:

Servers create `Session` objects and pass them to `Session_Manager::establish(Session) -> Session_Handle`. Hence, the `Session_Manager` is in full control of both the `Session_ID` and the `Session_Ticket`. Servers will retrieve `Session` objects directly via `Session_Manager::retrieve(Session_Handle) -> Session` (TLS 1.2) or via `Session_Manager::choose_from_offered_tickets`. Both are high-level methods implemented directly in the base class. They take care of session policy validation (e.g. expiry). Low-level storage methods must be implemented by all sub-classes.

Clients on the other hand always receive a `Session_Handle` from the server (that's just how TLS works). They use `Session_Manager::store(Session, Session_Handle)` to associate their sessions with the handles from the server. Usually, clients will then use `Session_Manager::find(Server_Information) -> std::vector<std::pair<Session, Session_Handle>>` to retrieve potential sessions for resumption. Again, `find` is a high-level method implemented in the base class in terms of a low-level abstract storage method.

Note that clients may retrieve multiple sessions for resumption. TLS 1.3 allows clients to provide several tickets (as PSK identities) at once. The `Session_Manager` can opt to always return a single `Session`, `Session_Handle` pair, of course. On the server-side `Session_Manager::choose_from_offered_tickets()` receives the tickets offered by the client and selects the "most favorable": The default implementation simply selects the first that matches the server's needs. 

I'm hoping that typical users won't need to override the high-level (policy-enforcing) methods. They are marked as `virtual` though and can be overridden if needed.

Session deletion (e.g. after usage by the client/server) is done with `Session_Manager::remove(Session_Handle)`.

### `class Session_Manager_Stateless`

A strawman manager for a stateless server. It generates random Session IDs and creates tickets via `Session::encrypt()`. The sessions are not persisted and `Session_Manager::retrieve(Session_ID)` will always fail. To stay compatible with the current architecture `Session_Manager_Stateless` uses `Credentials_Manager::psk()` to obtain the necessary ticket key.

### `class Session_Manager_In_Memory`

Like before, this keeps sessions in-memory in a `std::map<>`. Note that the `SessionID` strong-type acts as a key! This implementation will never emit session tickets.

### `class Session_Manager_SQLite`

I adapted the implementation to the new `Session_Manager` API. Most notably, the implementation now detects old (and/or corrupted) databases and resets them transparently. Some more clean-ups could be done regarding the SQL database and statements wrapper. As the size of this patch is already getting out of hand, that will come later.

### `class Session_Manager_Hybrid`

This internally combines a `Session_Manager_Stateless` with an arbitrary (stateful) `Session_Manager` implementation. It automatically creates either IDs or tickets depending on whether the TLS 1.2 peer can handle tickets. Applications can choose whether to prefer tickets or IDs at construction time. This is likely the most useful flavor for TLS servers.

## What's next

First: Thanks for reading through those thoughts. 😄

A few open questions:

* ~~Strong types: yay? or nay?~~
* ~~Should we go as far and separate the `Session_Manager` into a server and a client part?~~ Now that the API seems almost fully disjoint (apart from `::remove()`).
* ~~Required technical understanding of the application developer: is this taking it too far?~~
  * Generation of valid SessionIDs and Tickets
  * Lifetime handling
  * Choosing from a number of offered tickets
  * Balancing how many tickets a TLS 1.3 client should offer at once
  * ...
* ...

### CODE REVIEW

File: src/lib/utils/concepts.h
Comment (randombit) [inline src/lib/utils/concepts.h:30]: Re compiler guarantees, unlike GCC, Clang and MSVC where we draw a line in the sand and say "Version X is/will be supported", for XCode and NDK we just commit to supporting the latest version. So we could fix this at any point in the 3.x releases, assuming XCode and NDK get themselves sorted out eventually.

**Comment by randombit** [comment]
> Strong types: yay? or nay?

I haven't seen this before in C++ tbh, but I'm generally a fan of [newtypes in Rust](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) which seem to be basically the same idea. Fine to proceed along these lines.

> Should we go as far and separate the Session_Manager into a server and a client part. Now that the API seems almost fully disjoint (apart from ::remove()).

The underlying storage mechanism is going to be the same though, and we'd be duplicating that if we split the managers. 
(Possibly the real fix here is to make the  storage mechanism distinct from the policy logic)

> Required technical understanding of the application developer: is this taking it too far? 
> ...

I think we should provide toggles where someone might reasonably have differing needs, but in general people just want something to work and they shouldn't have to understand what any of this is doing unless they are doing something tricky or unusual. If in doubt we should just use whatever default parameters that say BoringSSL or OpenSSL uses.

File: src/lib/tls/tls12/tls_client_impl_12.cpp
Comment (reneme) [inline src/lib/tls/tls12/tls_client_impl_12.cpp:790]: This fixes some BoGo tests that were disabled with "needs investigation". Basically, ticket renewal was not implemented up until now.

**Comment by reneme** [comment]
This seems to finally converge. Frankly, I underestimated the intricacies of consolidating 1.2 and 1.3 behind the same API, but I'm fairly happy with this now. Commit history will need cleanup and potentially, I'll also split this into two or three patch sets.

Main improvements from an (advanced) user's perspective:

* Full control over the usage of stateful of stateless resumption by subclassing `Session_Manager`.
  * Previously, ticket creation was hard-coded
  * Stateful and stateless operation code is easily reusable for TLS 1.2 and 1.3
* A number of default `Session_Manager` classes to choose from:
  *  `..._In_Memory`
    as before, keeps state in RAM. When used in servers, it effectively disables stateless ticket support. When used in clients, tickets as well as IDs can be stored and resumed
  * `..._Stateless`
    effectively disables stateful resumption (does not store anything) in servers. Is not useful for clients.
  * `..._Hybrid`
    internally combines a `Session_Manager_Stateless` with an arbitrary (stateful) `Session_Manager` implementation. Automatically creates either IDs or tickets depending on whether the TLS 1.2 peer can handle tickets. Applications can choose whether to prefer tickets or IDs at construction time.

## Still to do

* [ ] Cleanup
* [ ] Move SQL-based managers to the new architecture
* [ ] Remove `Session_ID` and `Session_Ticket` from the `Session` object
   (Sessions don't need to be aware of their handle)
* [ ] Rebase #3140 onto this
    (I smoke-tested this locally already to validate the API)

## Questions

* The SQL database schema will likely change, do we need a migration for this?
* The Session DER-schema will likely change. Migration needed?
* Anything we could improve on the above schematics once we're on it?
* Do you think the documentation for `Session_Manager` is sufficient to enable users to custom-make their own?

**Comment by randombit** [comment]
> The SQL database schema will likely change, do we need a migration for this?
> The Session DER-schema will likely change. Migration needed?

IMO no. TLS sessions are transitory. We should just make sure it's clearly called out in the release notes and migration guide.

**Comment by reneme** [comment]
Sorry for leaving this dormant for a while. I waited for a continuous time slot to finish this up as it required some more focused refactoring and RFC pondering. Anyway, this should be the last oversized PR for TLS 1.3 😄.

Turns out that @randombit was right, that storage and policy should be somewhat separated:

>> Should we go as far and separate the Session_Manager into a server and a client part. Now that the API seems almost fully disjoint (apart from ::remove()).

> The underlying storage mechanism is going to be the same though, and we'd be duplicating that if we split the managers.
(Possibly the real fix here is to make the storage mechanism distinct from the policy logic)

I didn't split the managers into a client and a server part. Instead I added a layer of indirection in the `Session_Manager` API that allows implementing policy filters (mainly expiry) in the base class where all derived managers can 
benefit from it. Details are in Doxygen of `session_manager.h`.

#### Final touches

... that might be done in follow-ups as this PR is becoming ridiculously hard to reason about.

* [ ] Commit history cleanup
* [ ] Distinction between `Policy::session_ticket_lifetime()` and `Policy::tls12_session_lifetime()` (would be a new policy)
* [ ] `Session_Manager::used_in_handshake(bool success)` to inform the manager that a ticket was used in a handshake so that it may be removed

File: src/lib/tls/tls_session_manager.cpp
Comment (reneme) [inline src/lib/tls/tls_session_manager.cpp:68]: This applies the same expiry filter to both tickets and TLS 1.2 session IDs. Previously, there were two lifetimes:
* `Policy::session_ticket_lifetime()` and
* `Session_Manager_Memory/SQL(..., session_lifetime)`.

TLS 1.3 doesn't make this distinction anyway, but we could introduce a `Policy::tls12_session_id_lifetime()` to allow the same flexibility as before.

Note that previously derived `Session_Manager` classes each needed to implement the expiry logic. That is still possible (by overriding `::retrieve()` and `::find()`) but not necessary anymore. For applications to use the default implementation, they would just need to override the "internal storage methods" `::retrieve_one()` and `::find_all()`.

File: src/lib/tls/sessions_sql/tls_session_manager_sql.cpp
Comment (reneme) [inline src/lib/tls/sessions_sql/tls_session_manager_sql.cpp:48]: SQL database migration is "lazy" in a sense that it removes the old tables and recreated new ones with an updated schema. The same mechanism is used when the database is found to be corrupted.

File: src/lib/tls/tls_session_manager.cpp
Comment (reneme) [inline src/lib/tls/tls_session_manager.cpp:167]: Now that this base-implementation of "single-use tickets" utilizes the fact that TLS implementations use the `find()` method atomically we should guard the base implementation with a mutex.

This should be independent of the thread safety requirement of the derived managers, though. Their overrides might be called directly, and hence need to be mutexed as well. However, we could provide a protected interface to reuse the base-classes mutex.

- [ ] Make `Session_Manager` base class thread safe

File: src/lib/tls/tls_client.cpp
Comment (reneme) [reply src/lib/tls/tls_client.cpp]: No more! Instead #3244 refactors this part of  the code to ensure that `find()` is only called once for each handshake. I.e. `Session_Manager` can implement single-use tickets by atomically removing `Session` objects it hands out to the TLS stack.

**Comment by reneme** [comment]
A side note on thread safety: The `Session_Manager_SQL` did not use a mutex so far. I believe it should though. I guess the implementation relied on SQLite being thread-safe by default. Though, this can be disabled by the application developer (e.g. when compiling their SQLite in single-threaded mode). Also, custom SQL adapters might not provide the same guarantees.

- [ ] consider mutexing method invocations in `Session_Manager_SQL` using `lock_guard_type lk(*this)`

File: src/lib/tls/tls_session_manager.cpp
Comment (reneme) [reply src/lib/tls/tls_session_manager.cpp:175]: Initially, I wanted to make `Session_Manager` implement the C++11 `Lockable` requirement and let derived classes reuse the base class' mutex simply by: `Botan::lock_guard_type lk(*this)`. Unfortunately, CTAD for aliased templates (`Botan::lock_guard_type`) introduced in C++20 doesn't seem to be implemented by clang (yet?). So one would have explicitly needed to state `Botan::lock_guard_type<Session_Manager>`.

All of that felt like magic and eventually I just exposed `Session_Manager::m_mutex` via a protected accessor for derived classes to use.

**Comment by reneme** [comment]
rebased to master to solve conflicts with enum adaptions.

**Review by lieser — COMMENTED** [review]
First incomplete round of review.

File: src/tests/test_tls_session_manager.cpp
Comment (lieser) [inline src/tests/test_tls_session_manager.cpp:204]: From my understanding it never contains a ticket, so the "dropped" is a little confusing to me.

File: src/tests/test_tls_session_manager.cpp
Comment (lieser) [inline src/tests/test_tls_session_manager.cpp:228]: The test description alone sounds like it should work, even if it does in fact not work.

File: src/tests/test_tls_session_manager.cpp
Comment (lieser) [inline src/tests/test_tls_session_manager.cpp:870]: Why do we need the sleep? Same below.

File: src/tests/test_tls_session_manager.cpp
Comment (lieser) [inline src/tests/test_tls_session_manager.cpp:972]: Do we also somewhere test that the reuse of sessions works if the policy allows it?

File: src/lib/tls/sessions_sql/tls_session_manager_sql.cpp
Comment (reneme) [reply src/lib/tls/sessions_sql/tls_session_manager_sql.cpp:219]: True! That's a left-over from a previous API approach. Good catch!

File: src/lib/tls/tls12/tls_client_impl_12.cpp
Comment (reneme) [reply src/lib/tls/tls12/tls_client_impl_12.cpp:806]: It's being tackled with the implementation of [this comment](https://github.com/randombit/botan/pull/2988#issuecomment-1351250465). (we have a ticket for this in our internal Jira)

File: src/lib/tls/msg_client_hello.cpp
Comment (reneme) [reply src/lib/tls/msg_client_hello.cpp:358]: Good catch! If I may add: It should then also be non-const to (hopefully) let the compiler move it into the return value.

File: src/lib/tls/tls_session_manager.h
Comment (reneme) [reply src/lib/tls/tls_session_manager.h:114]: For consistency I agree. But on the other hand its actually quite handy for functions that may return different things depending on the situation.

Quoting the Doxygen manual:

> Starts a description for a function's return value with name <return value>, followed by a description of the return value. The text of the paragraph that forms the description has no special internal structure. All visual enhancement commands may be used inside the paragraph. Multiple adjacent \retval commands will be joined into a single paragraph. Each return value description will start on a new line.

This is how the rendering looks like:

image

... and I actually agree: its more for "multiple" return values than for different value alternatives. Changing...

File: src/lib/tls/tls_session_manager.h
Comment (reneme) [reply src/lib/tls/tls_session_manager.h:204]: Certainly.

File: src/lib/tls/tls_session_manager.h
Comment (reneme) [reply src/lib/tls/tls_session_manager.h:262]: Mhm, that's a fair point. However, the entire TLS stack works under the assumption that the RNG object lives as long as the TLS session is active:
The top-level `TLS::Server` (just like client) c'tor takes an `RNG&` and stores it as a reference (e.g. in `TLS::Channel_Impl_12`). The RNG is then used throughout by passing it around.

That said: I believe all the other TLS classes don't retain the reference but rely on the caller to pass the RNG reference where needed. That is something we could consider for the `Session_Manager`.

Leaving this open for now.

File: src/lib/tls/tls_session_manager.cpp
Comment (reneme) [reply src/lib/tls/tls_session_manager.cpp:158]: Nothing to consider here. What was I thinking. 😨

File: src/lib/tls/tls_session_manager.cpp
Comment (reneme) [reply src/lib/tls/tls_session_manager.cpp:27]: I tend to agree. Let's spike an alternative API together.

File: src/lib/tls/tls_session_manager_stateless.cpp
Comment (reneme) [reply src/lib/tls/tls_session_manager_stateless.cpp:61]: In TLS 1.3 PSK and resumption (aka session handling) are one and the same mechanism. That's why PSK shows up here. Let me quote more context from the RFC:

> Implementor's note: When session resumption is the primary use case
   of PSKs, the most straightforward way to implement the PSK/cipher
   suite matching requirements is to negotiate the cipher suite first
   and then exclude any incompatible PSKs.  Any unknown PSKs (e.g., ones
   not in the PSK database or encrypted with an unknown key) SHOULD
   simply be ignored.  If no acceptable PSKs are found, the server
   SHOULD perform a non-PSK handshake if possible.  If backward
   compatibility is important, client-provided, externally established
   PSKs SHOULD influence cipher suite selection.

Comment (reneme) [reply src/lib/tls/tls_session_manager_stateless.cpp:61]: If we fail to decrypt a received ticket we simply ignore that and proceed to a full handshake. (That's what happens eventually, when this method returns `std::nullopt`)

File: src/lib/tls/tls_session_manager_stateless.cpp
Comment (reneme) [reply src/lib/tls/tls_session_manager_stateless.cpp:75]: I added a docstring to the constructor of this class that reads:

```C++
/**
 * The key to encrypt and authenticate session information will be drawn
 * from @p credentials_manager as `psk("tls-server", "session-ticket")`.
 * It is the responsibility of the calling application to set up its own
 * Credentials_Manager to provide a suitable key for this purpose.
 */
```

File: src/lib/tls/tls_session_manager_hybrid.h
Comment (reneme) [reply src/lib/tls/tls_session_manager_hybrid.h:70]: Its fine to use `Session_Manager_Hybrid` with a suitable storage backend (e.g. `Session_Manager_In_Memory`) for a client as well. The "stateless" path of the hybrid manager would never actually be exercised, though. Albeit, on the client side, it doesn't make a difference whether someone uses `SM._Hybrid` or `SM._In_Memory`.

File: src/tests/test_tls_session_manager.cpp
Comment (reneme) [reply src/tests/test_tls_session_manager.cpp:204]: Mhm, that's a left-over from a previous revision. At one point I thought the `Session_Handle` could host both an ID and a Ticket. I later understood that this doesn't really make sense. I'll change the label to "not a ticket".

File: src/tests/test_tls_session_manager.cpp
Comment (reneme) [reply src/tests/test_tls_session_manager.cpp:228]: Indeed. Thanks. Similar excuse as above. The assumptions changed on the way and the test names didn't keep up.

File: src/lib/tls/tls_session_manager.cpp
Comment (reneme) [reply src/lib/tls/tls_session_manager.cpp:158]: ... turns out: This is done because `resize()` requires the members of the vector to be default-constructible. After all it might also be used to extend the vector.

File: src/tests/test_tls_session_manager.cpp
Comment (reneme) [reply src/tests/test_tls_session_manager.cpp:972]: Not explicitly. Though, some tests above rely on this policy being `true` by default.

**Comment by reneme** [comment]
Thanks @lieser for the thorough review! You certainly found quite a few places with potential for improvement.

Not yet addressed are (IMHO):

* [ ] https://github.com/randombit/botan/pull/3150#discussion_r1112201037
* [ ] https://github.com/randombit/botan/pull/3150#discussion_r1114403042

File: src/lib/tls/tls_session_manager.h
Comment (lieser) [reply src/lib/tls/tls_session_manager.h]: I thought it was just an alias and didn't know it had a different meaning. So I think it would be ok to start using it then it makes sense.

File: src/lib/tls/tls_session_manager.h
Comment (randombit) [reply src/lib/tls/tls_session_manager.h:262]: One thing we could consider is something like:

`std::unique_ptr<RandomNumberGenerator> RandomNumberGenerator::fork()`

where for RNGs which are internally a singleton (Processor_RNG, System_RNG, etc) it just returns a new instance of itself in a `unique_ptr`, or for RNGs like HMAC_DRBG, returns a new instance that is seeded with 256 bits taken from the source RNG.

**Review by randombit — APPROVED** [review]
Did not do a terribly careful review here, but leaving an approval to indicate I didn't notice any immediate problems / no objections to merging.

File: src/lib/tls/tls_session_manager.h
Comment (reneme) [reply src/lib/tls/tls_session_manager.h:262]: One more thing to consider: The `Session_Manager`, unlike other (internal) TLS classes, is ultimately managed and owned by the application code. I.e. the `Session_Manager` itself must be kept alive for the lifetime of the TLS server/client. Changing those lifetime expectations just for the `Session_Manager` might be more confusing than helpful.

@randombit you are basically proposing that the TLS stack as a whole gets it's own forked-off instance of the RNG. Or do you mean that both the TLS stack and the `Session_Manager` each would get independent RNG instances?

In any case: For the sake of scope, I suggest to take this on after this PR-chain is done and merged.

Comment (reneme) [reply src/lib/tls/tls_session_manager.h:262]: Follow up: https://github.com/randombit/botan/issues/3321

File: src/lib/tls/tls12/tls_channel_impl_12.cpp
Comment (reneme) [reply src/lib/tls/tls12/tls_channel_impl_12.cpp]: Yeah. Initially, I thought that this `std::pair<>` would be fairly local. It got to a point where a dedicated aggregator type is certainly justified. See here: https://github.com/randombit/botan/pull/3150/commits/f9637772da42fa0daa29d0a2f2219c593acb238d

**Comment by reneme** [comment]
Done from my side. Waiting for CI and @lieser's approval.

**Review by lieser — APPROVED** [review]

**Comment by reneme** [comment]
# 🥳

**Comment by oviano** [comment]
> A side note on thread safety: The Session_Manager_SQL did not use a mutex so far. I believe it should though. I guess the implementation relied on SQLite being thread-safe by default. Though, this can be disabled by the application developer (e.g. when compiling their SQLite in single-threaded mode). Also, custom SQL adapters might not provide the same guarantees.

I wanted to comment on this as one of my use/test cases is a client-side test application which opens tens of thousands of TLS connections to a server. To force it to store separate session records for each client, despite connecting to the same host and port, I plan to append an id to the host name prior to the look-up.

It seems unfortunate to me not to be able to leverage the multi-threading capbilities of SQLite, and that we essentially have to serialise these look-ups.

Under 1.2 I had written my own session manager, but I figured with all the changes to the API, rather than retrofit what I have I'd give the SQLite one a go.

Is it really necessary to lock a mutex around every single SQL look-up?

Perhaps it can be optional as a template argument to the session manager allowing you to supply a no-op mutex or something...

**Comment by reneme** [comment]
> Is it really necessary to lock a mutex around every single SQL look-up?

I see this mainly as a safety precaution. But perhaps you are right and that's a bit too broad. Maybe we could use [`sqlite3_threadsafe()`](https://www.sqlite.org/c3ref/threadsafe.html) to figure out whether we need the mutex or not?

**Comment by oviano** [comment]
Reading up a little bit on SQLite, my understanding is that there are three settings.

1. Single threaded, so no mutex, not thread safe at all.
2. Multi-threaded, allowing different threads to access the DB at the same time, provided they use their own connection
3. Serialized, so SQLlite allows full access to the same database connection from multiple threads and applies the necessary locking to achieve this.

Actually, in my use case, where I have a fixed number of worker threads servicing connections then it would seem that something could be possible with option 2 above in conjunction with a thread_local helper to manage the database connection lifetime.

Maybe it would be a custom/derived SQLite Session_Manager, but there is still some locking going on in the base Session_Manager and Session_Manager_SQL classes which would need addressing somehow...so maybe something along the lines of being able to provide the mutex type (or a no-op) for the Session_Manager base class as a template argument and then the current Session_Manager_SQLite would pass in recursive_mutex_type, and some new Session_Manager_SQLite_TL (thread local) would pass in the noop_mutex but then would change how the DB is created, to do some sort of thread local magic to ensure one database connection exists per thread, plus also specifying to sqlite3_open that it can operate in multithreaded mode.

Just some thoughts anyway.

**Pull request merged**
